# pacific_albacore_ddRADseq
Commands and input files for population genetic analysis of Pacific Albacore using ddRADseq

# albacore_STACKS_and_filtering
Lists input commands used for STACKS and BWA, and subsequent filtering in VCFtools, plink, and paralog-finder (HDplot).

# maps
Maps used to select individuals and populations in STACKS runs.

# excluded_loci
Lists used to exclude certain loci in STACKS runs.

# plink-ld
Ped and map files used by plink to estimate loci in LD.

# vcftools-hwe
Maps used to select subsets of individuals for estimating HWP.

# R
R files for running loci filtering and population genetic analyses (e.g. whoa, adegenet, fsthet). I recommmend viewing original source GitHub pages and tutorials noted in each R file credits.

# Sampling, and genotype and loci fasta files
Sampling and sequence spreadsheets, genetic data (vcf, genepop, and structure genotype files, and the fasta consensus file for the 84 putatively adaptive loci), and full results for genome scan program and Fst analyses are available from Dryad: https://doi.org/10.5061/dryad.6djh9w103

# Demultiplexed reads
Demultiplexed forward and reverse DNA sequence reads for the albacore sequenced in this study are openly available on the NCBI sequence read archive (SRA) under: PRJNA579774, http://www.ncbi.nlm.nih.gov/bioproject/579774

# Raw Illumina reads
Available on request from Kathleen O'Malley (State Fisheries Genomics Lab, Oregon State University), https://agsci.oregonstate.edu/state-fisheries-lab/people

# Further information
See this page: https://sites.google.com/view/evauxlution/data
